### DISCUSSION TITLE
Rollup is creating a new `useStore` for every module

### DISCUSSION THREAD

**Discussion by kylemilloy** [pmndrs/zustand#2308] [Q&A] [answered]
I'm wondering if anyone else has encountered this or has some more experience than I in getting rollup to do what I want it to.

We have a set up at work where we have some shared files that get processed by rollup and exported as several modules that we can import into other packages in our monorepo. What ends up happening is that if I have two separate modules that are importing the same `useStore` in my typescript, after I run rollup I get two modules, each with their own `const useStore = zustand.create(persistedStore())`, and it results in state not being consistent. For example. Before roll up I'll have something like this:

```tsx
// store module
const persistedStore = () => persist((...args) => ({
  ...createAuthSlice(...args)
}, { name: 'storage-key' })

export const useStore = create(persistedStore())

// auth module
import { useStore } from '@/store'

export function useAuth() {
  const [permissions, setPermissions] = useStore((state) => [state.permissions, state.setPermissions])
  const can = useCallback((permission) => permissions.includes(permission), [permissions])  

  return {
    can,
    setPermissions
  }
}

// checkout module
import { useStore } from '@/store'

export function useCheckout() {
  const { can } = useAuth()

  const purchase = useCallback(() => {
    if (!can('make-purchase')) {
      throw new Error('unable to purchase')
    }

    // carry on
  }, [can])

  return { purchase }
}
```

So these are two example hooks I would have. Let's say on a log in page I have the following component:

```tsx
import useAuth from '@/dist/modules/auth'
import useCheckout from '@/dist/modules/checkout'

export default function LoginPage() {
  const { can, setPermissions } = useAuth()
  const { purchase } = useCheckout()

  return (
    <>
      <button onClick={() => setPermissions(['make-purchase'])}>Login</button>
      {can('make-purchase') && <button onClick={purchase}>Purchase</button>}
    </>
  )
}
```

So what I am seeing in my component is that in this example the `can` method in useAuth works as intended: When I 'log in' it sets the permissions and then it shows the button to make a purchase because its version of `can()` has the updated permissions. HOWEVER when I click purchase it always throws as the version in useCheckout does not contain this information. When I go examine my rollup output I see that I have two files that look like this:

```js
// modules/store.js
// ...all the slices/middleware/etc.
const useStore = zustand.create(persistedStore())
exports.useStore = useStore

// modules/auth.js
const useStore = zustand.create(persistedStore())

function useAuth() {
  const [permissions, setPermissions] = useStore((state) => [state.permissions, state.setPermissions])
  const can = useCallback((permission) => permissions.includes(permission), [permissions])  

  return {
    can,
    setPermissions
  }
}
exports.useAuth = useAuth

// modules/checkout.js
const useStore = zustand.create(persistedStore())

function useAuth() {
  // ... all the user auth stuff.
}

function useCheckout() {
  const { can } = useAuth()

  const purchase = useCallback(() => {
    if (!can('make-purchase')) {
      throw new Error('unable to purchase')
    }

    // carry on
  }, [can])

  return { purchase }
}

exports.useCheckout = useCheckout
```

I suspect the problem here again is that it instantiates a new `useStore` for each hook instead of importing it from the file. I see this behaviour in common and in the esm. Not sure if there's a way to control this behaviour so it's doing something like:

```js
const useStore = require('./modules/store')
// or in esm
import useStore from './modules/store'

function useAuth() {
  const permissions = useStore(state => state.permissions)

  // ...
}
```

Any help would be appreciated. Also, lemme know if this would be better posted in the rollup repo but I figured I'd ask here to test my hunch RE: rollup spitting out multiple instances of `useStore`

**Comment by dbritto-dev** [comment]
@kylemilloy would you mind sharing us you rollup conf? btw, are you using vite or just pure rollup?

**Comment by kylemilloy** [comment]
Hey thanks for following up @dbritto-dev . Just rollup, no vite. The conf is a doozy...see below.

```js
import dts from "rollup-plugin-dts";
import esbuild from "rollup-plugin-esbuild";
import alias from "@rollup/plugin-alias";
import hq from "alias-hq";
import autoExternal from "rollup-plugin-auto-external";
import resolve from "@rollup/plugin-node-resolve";
import replace from "@rollup/plugin-replace";
import rollupEnv from "./rollup-env";
import modules from "./modules";

const aliasConfig = {
  customResolver: resolve({ extensions: [".tsx", ".ts"] }),
  entries: hq.get("rollup", { format: "array" }),
};

const plugins = [
  alias(aliasConfig),
  esbuild({
    target: "esnext", // let app bundlers compile to es5 target
  }),
  replace({
    ...rollupEnv,
    preventAssignment: true,
  }),
  autoExternal(),
];

const dtsPlugins = [dts()];

const toExport = [];
modules.forEach((module) => {
  toExport.push({
    input: module.inputFile,
    output: [
      {
        file: `dist${module.outputDir}${module.name}.js`,
        format: "cjs",
        sourcemap: true,
      },
      {
        file: `dist${module.outputDir}${module.name}.esm.js`,
        format: "esm",
        sourcemap: true,
      },
    ],
    plugins,
    // Treat the dependency as external and prevent Rollup
    // from trying to include it in the bundle.
    // Suppresses https://rollupjs.org/troubleshooting/#warning-treating-module-as-external-dependency
    external: [
      "react/jsx-runtime",
      "lodash/isObject",
      "lodash/get",
      "lodash/assign",
      "lodash/reject",
      "lodash/filter",
      "lodash/sumBy",
      "lodash/uniq",
      "lodash/find",
      "lodash/values",
      "lodash/clone",
      "lodash/map",
      "lodash/chain",
      "lodash/each",
      "lodash/throttle",
      "lodash/cloneDeep",
      "lodash/set",
      "lodash/isEqual",
      "lodash/omit",
      "lodash/debounce",
      "next/router",
      "country-region-data/data.json",
      "next/dynamic",
      "react-query/devtools",
      "react-query/devtools/development",
      "text-clipper",
      "konva/lib/Core",
    ],
  });

  // Compiles d.ts files in ./types to the proper module file
  // in dist/[moduleName].d.ts.
  // We use tsc to compile the ts and tsx files to d.ts and then
  // when rollup compiles the modules we use the below to grab the
  // correct d.ts file location (since the module folder structure
  // may not necessarily match the original typescript file structure)
  const pathStart = "./";
  const inputFile = module.inputFile.substring(
    pathStart.length,
    module.inputFile.indexOf(".ts")
  );

  toExport.push({
    input: `./types/${inputFile}.d.ts`,
    output: {
      file: `dist${module.outputDir}${module.name}.d.ts`,
      format: "esm",
    },
    plugins: dtsPlugins,
  });
});

export default toExport;

```

**Reply by dbritto-dev** [reply]
@kylemilloy I bet the issue is due to treating each file file as an entry

**Reply by kylemilloy** [reply]
Thanks for the insight @dbritto-dev . Currently our use is that imports are done as such...

```ts
import { useAuth } from '@shared/auth'
import { useCheckout } from '@shared/checkout'
```

...to separate things. If we made everything available via a single entry I am going to assume this would turn into something like...

```ts
const { useAuth, useCheckout } from '@shared'
```

...or is it possible to maintain the same paths as above but have it so not everything is an entry? I apologize if that's a stupid question but honestly I usually just leave the defaults on for whatever tooling I'm using in webpack/vite/rollup so this level of customization is a bit above me.

**Reply by dbritto-dev** [reply, marked as answer]
@kylemilloy let's say we have two entries but we want just one so both can be treated as one entry

It can be done like this:

```
// shared/auth.ts

export const authing = 'something'
```

```
// shared/checkout.ts

export const checkouthing = 'something'
```

```
// shared/index.ts

export * from './auth.ts'
export * from './checkout.ts'
```

Now we can have `shared/index.ts` as an entry instead of `shared/auth.ts` and/or `shared/checkout.ts`

> **Note:** If that doesn't separate the chunks you can manually specify the chunks, would be `auth` and `checkout`.

**Reply by kylemilloy** [reply]
Thanks @dbritto-dev 

I think as there is a significant amount of sunk time into the way we have rollup configured and little appetite to make significant changes, the approach I may take is to treat the store as an external import, build it separately, and then import into my shared hooks. It's a bit kludgey but at this time the quickest path forward for this.

Thank you again for your help and recommendations.

**Reply by dbritto-dev** [reply]
@kylemilloy yeah, that works too
